### ISSUE TITLE
request : adding a graph view like the one in Obsidian

### ISSUE THREAD

**Issue by mlaarebi** [pbek/QOwnNotes#2074]
#### Expected behaviour
To have a graph view that enables to see the relations between notes, just like the one in Obsidian.

#### Actual behaviour
Qownnotes doesn't have one, yet.

**Comment by pbek** [comment]
Can you please elaborate on your request.

**Comment by mlaarebi** [comment]
a graph view enables the user to see a map of all the connections between his linked notes. A map that he can interact with to visualize every connection between his notes. 

I uploaded a screenshot (as example) from the Obsidian app.

![Screenshot_2021-03-19 Obsidian A knowledge base that works on local Markdown files ](https://user-images.githubusercontent.com/50870076/111813434-7d83c300-88d9-11eb-8343-b547da468d88.png)

**Comment by pbek** [comment]
Thank you. What, in your (or Obsidian's) eyes, are linked notes? Markdown link to other note file? Same Tag? Notes with the same parent tag linked (how many levels up)? Same sub-folder?

**Comment by mlaarebi** [comment]
They are Markdown files linked by internal links (just like in Qownnotes), the difference is the "graph view" that can help to visualize the relations between notes.

Meaning that Qownnotes already have everything needed for this future, it only needs the ability to preview the links visually. Now we can preview the markdown files, It would be nice to add an option to show a graph of all the links between notes in a sub folder.

**Comment by pbek** [comment]
> linked by internal links

What are "internal links"? 😁 

The only information currently available are notes linked by tags. Are we talking about tags?

**Comment by mlaarebi** [comment]
you can link notes directly without tags, just by adding its link (the icon in the picture).

![Screenshot from 2021-03-19 22-21-25](https://user-images.githubusercontent.com/50870076/111843595-c51e4500-8901-11eb-8781-2aed867278a6.png)

**Comment by fabianski7** [comment]
[mindforger](https://github.com/dvorka/mindforger) has something similar.
https://mindforger.com/

**Comment by pbek** [comment]
> you can link notes directly without tags, just by adding its link (the icon in the picture).

Yes, of course you can. 😁  But QON does not store that information other than by putting a file link (some text) into the note text file. So linking information needs to be gathered (parsed and evaluated) for all notes.

> mindforger has something similar.

I can see a "tag navigator", are you talking about that, @fabianski7?
There are people who "link" their notes just by tags.

**Comment by pbek** [comment]
@mlaarebi, so what is in Obsidian a linked note? Or does obsidian have no tags?

**Comment by mlaarebi** [comment]
@pbek Obsidian uses tags too, but also links between them by internal links (like in Qownnotes).

And calls the "sub folders" : "vaults". 

Obsidian's linked note = Qownnotes's linked notes.
Obsidian's tags = Qownnotes's tags.

**Comment by fabianski7** [comment]
> I can see a "tag navigator", are you talking about that, @fabianski7?

no. it is called Knowledge Graph Navigator 
https://github.com/dvorka/mindforger-repository/blob/master/memory/mindforger/user-documentation.md#tayb-think-as-you-browse---knowledge-graph-navigator-

**Comment by pbek** [comment]
Thank you.

**Comment by jg55** [comment]
Just adding a friendly +1 to this feature request, this would indeed be a helpful feature. Obsidian's graph view is excellent.

**Comment by pbek** [comment]
It's a lot simpler to do in web apps. 😁

**Comment by mlaarebi** [comment]
> It's a lot simpler to do in web apps. grin

Obsidian is a desktop app (mobile too but in the future).

**Comment by pbek** [comment]
> Obsidian is a desktop app

Yes, but it looks like an Electron app (= web app) 😉  ...and closed source, so I can't look
Mindforger is a Qt app. 👍🏻

**Comment by gour** [comment]
Hello,

> Yes, but it looks like an Electron app (= web app) wink ...and closed source, so I can't look Mindforger is a Qt app. 👍🏻

I'm still shopping my "future-proof" note-taking application allowing me to use (neo)vim as external editor.

It would be nice to have graph-view and here is the Joplin [plugin](https://github.com/treymo/joplin-link-graph) which does implement it, if it helps?

**Comment by gour** [comment]
> Joplin is also a web-app sweat_smile

Right, but, at least, there is a source of the plugin available. :wink:

**Comment by SpyrosMourelatos** [comment]
Maybe a compromise  between difficulty in implementation and feature richness would be to make a DOT file [https://en.wikipedia.org/wiki/DOT_(graph_description_language)](https://en.wikipedia.org/wiki/DOT_(graph_description_language)) and then let the user visualize the graph manually. Then the only job for QownNotes would be to parse the markdown files and whenever a tag refers to another make a graph connection on the dot file

**Comment by pbek** [comment]
You can manipulate the preview with the scripting engine...

**Comment by MAAliy** [comment]
Is it possible to have something like this? https://github.com/treymo/joplin-link-graph

**Comment by MAAliy** [comment]
> > Is it possible to have something like this? https://github.com/treymo/joplin-link-graph
> 
> Are we talking about [#2074 (comment)](https://github.com/pbek/QOwnNotes/issues/2074#issuecomment-939284072)?

ah yes well how about trillium or siyuan? They are both offline.

**Comment by pbek** [comment]
Are those Qt apps?

**Comment by louwers** [comment]
Both are web apps. 

Graph drawing is pretty cool: https://en.wikipedia.org/wiki/Graph_drawing

Just adding to the stack of links here, this looks interesting: https://github.com/ogdf/ogdf

There is a QML binding available (9 years old :sweat_smile:) https://github.com/schulzch/qml-ogdf

Video here: https://www.youtube.com/watch?v=ZOw1eOYFumo&t=7s

**Comment by pbek** [comment]
> Both are web apps.

😅 

> There is a QML binding available (9 years old sweat_smile) https://github.com/schulzch/qml-ogdf ([fork](https://github.com/BioDataAnalysis/qml-ogdf) from 2017)

Looks like this would very hard to get running on all platforms and build systems... and to maintain!

**Comment by Beurt** [comment]
> Maybe a compromise between difficulty in implementation and feature richness would be to make a DOT file https://en.wikipedia.org/wiki/DOT_(graph_description_language) and then let the user visualize the graph manually. Then the only job for QownNotes would be to parse the markdown files and whenever a tag refers to another note make a graph connection on the dot file

There is already a QON plugin to render DOT files: <https://github.com/qownnotes/scripts/tree/master/render-plantuml>. So, if someone can manage to generate a DOT file from `.md` files (a map of all the links inside the files), the rest is already done.

**Comment by Beurt** [comment]
@SpyrosMourelatos @pbek 

I'm adding that there is already something that could help build the DOT graph of connections for a specific note. There is two scripts: [One for wikilinks support `[[notename]]`](https://github.com/qownnotes/scripts/tree/master/wiki-links) and [another for listing the backlinks using such format](https://github.com/qownnotes/scripts/tree/master/backlinks). The later output could be used to generate a DOT graph. However, the links between notes must use wikilinks (I think that proper MarkDown links `[text](URL)` are harder to use because the URL can use many paths such as relative, absolute, softlinks, etc.)

**Comment by blacklightpy** [comment]
The way to do this would be to create a database of all links by recursively parsing all notes.
Every link would count as a forward link.
Every page that is linked to would list the pages that link to them as backlinks.

Additionally, when using WikiLinks, anchor points could be supported to point to specific headings, or if there is a concept such as blocks in QOwnNotes. Or it could simply define a random marker (using an alphanumeric code by default, such as `^xyzz12`. The links would be like [[note-filename^xyzz12]].

And we already have custom names in Wikilinks, which we can pair with it like [[note-filename^xyzz12|Note Name]].

**Comment by blacklightpy** [comment]
GraphViz also has an [sfdp](https://graphviz.org/docs/layouts/sfdp/) (scalable force directed placement) layout engine.

There's also Qt [Elastic Nodes example](https://doc.qt.io/qt-6/qtwidgets-graphicsview-elasticnodes-example.html) based on QGraphicsScene.

**Comment by pbek** [comment]
## 25.3.0
- There now is a new **note relations panel**, that shows links to other notes in a graph
  for the current note, up to the 3rd level (for [#2074](https://github.com/pbek/QOwnNotes/issues/2074))
  - By double-clicking on a note in the graph, you can open the note

It was a lot of work, but I guess it still needs more work.

**Comment by pbek** [comment]
There now is a new release, could you please test it and report if it works for you?

**Comment by blacklightpy** [comment]
This makes me realize Flatpaks are the way to go for app distribution, because my distro didn't update this yet.

**Comment by pbek** [comment]
Oh, I have packages for lots of distributions.. 😉 But flatpaks are nice.

**Comment by pbek** [comment]
## 25.3.3
- You can now **zoom** in the **note relation panel** with the mouse wheel
  (for [#2074](https://github.com/pbek/QOwnNotes/issues/2074))

## 25.3.2
- In the **note relation panel**, notes can now have **circular relations**
  between each other, no new note items will be drawn for notes that are already
  in the scene (for [#2074](https://github.com/pbek/QOwnNotes/issues/2074))

## 25.3.1
- In the **note relation panel**, the gathering of note relations will now be done
  in a **different thread** (for [#2074](https://github.com/pbek/QOwnNotes/issues/2074))
  - This results in a huge speed improvement and being able to interact with the
    user interface while the relations are gathered and drawn
  - The gathering of note relations and drawing is stopped when another note is selected

**Comment by pbek** [comment]
There now is a new release, could you please test it and report if it works for you?

**Comment by 5472qaywsx** [comment]
Works for me like a charm, it resolved one issue I noticed in 25.3.1 with "backlinks" (notes containing links to one another, you called it "circular relations"): previous version 25.3.2 visualized the notes twice in the graph, now it's only once.

Before:

![Image](https://github.com/user-attachments/assets/cc44a318-f188-4f6f-9ecc-d62fc34484dc)

After:

![Image](https://github.com/user-attachments/assets/38bb53aa-d957-4dbc-8a15-cf91d1ab47df)

**Comment by pbek** [comment]
Great, thank you for testing!

**Comment by pbek** [comment]
I added a documentation page on https://www.qownnotes.org/getting-started/note-relations.html.
